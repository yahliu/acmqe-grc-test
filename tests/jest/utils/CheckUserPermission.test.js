/* Copyright (c) 2021 Red Hat, Inc. */
/* Copyright Contributors to the Open Cluster Management project */

'use strict'

import {
  checkCreatePermission,
  checkEditPermission,
  checkCreateRole,
  checkEditRole
} from '../../../src-web/utils/CheckUserPermission'

const accessUsers = {
  clusterAdmin: [{
    namespace: 'default',
    rules: {
      '*/*': ['*']
    }
  }],
  admin: [{
    namespace: 'default',
    rules: {
      'policy.open-cluster-management.io/policies': [
        'get',
        'list',
        'watch',
        'update',
        'create',
      ]
    }
  }],
  edit: [{
    namespace: 'default',
    rules: {
      'policy.open-cluster-management.io/policies': [
        'get',
        'list',
        'watch',
        'update',
      ]
    }
  }],
  view: [{
    namespace: 'default',
    rules: {
      'policy.open-cluster-management.io/policies': [
        'get',
        'list',
        'watch',
      ]
    }
  }]
}

const extraPolicyKeys = [ 'policy.open-cluster-management.io/policyautomations' ]

const mixedUsers = {
  mixedNsAdminEdit: [{
    namespace: 'namespace-1',
    rules: {
      'policy.open-cluster-management.io/policies': [
        'get',
        'list',
        'watch',
        'update',
        'create',
      ]
    }
  },{
    namespace: 'namespace-2',
    rules: {
      'policy.open-cluster-management.io/policies': [
        'get',
        'list',
        'watch',
        'update',
      ]
    }
  }],
  mixedKindAdminEdit: [{
    namespace: 'namespace-1',
    rules: {
      'policy.open-cluster-management.io/policies': [
        'get',
        'list',
        'watch',
        'update',
        'create',
      ],
      'policy.open-cluster-management.io/policyautomations': [
        'get',
        'list',
        'watch',
        'update',
      ]
    }
  }],
  mixedNsEditView: [{
    namespace: 'namespace-1',
    rules: {
      'policy.open-cluster-management.io/policies': [
        'get',
        'list',
        'watch',
        'update',
      ]
    }
  },{
    namespace: 'namespace-2',
    rules: {
      'policy.open-cluster-management.io/policies': [
        'get',
        'list',
        'watch',
      ]
    }
  }],
  mixedKindEditView: [{
    namespace: 'namespace-1',
    rules: {
      'policy.open-cluster-management.io/policies': [
        'get',
        'list',
        'watch',
        'update',
      ],
      'policy.open-cluster-management.io/policyautomations': [
        'get',
        'list',
        'watch',
      ]
    }
  }],
}

describe('checkCreatePermission', () => {
  for (const user in accessUsers) {
    it(`should return correct permissions for ${user} users`, () => {
      expect(checkCreatePermission(accessUsers[user])).toMatchSnapshot()
    })
  }
  it('should return 1 for a user with admin and edit in diff ns', () => {
    expect(checkCreatePermission(mixedUsers.mixedNsAdminEdit)).toMatchSnapshot()
  })
  it('should return 0 for a user with edit and view in diff ns', () => {
    expect(checkCreatePermission(mixedUsers.mixedNsEditView)).toMatchSnapshot()
  })
})

describe('checkCreateRole', () => {
  for (const user in accessUsers) {
    it(`should return correct permissions for ${user} users`, () => {
      expect(checkCreateRole(accessUsers[user][0].rules)).toMatchSnapshot()
    })
  }
  it('should return 0 for a user with admin and edit in diff kinds', () => {
    expect(checkCreateRole(mixedUsers.mixedKindAdminEdit[0].rules, extraPolicyKeys)).toMatchSnapshot()
  })
  it('should return 0 for a user with edit and view in diff kinds', () => {
    expect(checkCreateRole(mixedUsers.mixedKindEditView[0].rules, extraPolicyKeys)).toMatchSnapshot()
  })
})

describe('checkEditPermission', () => {
  for (const user in accessUsers) {
    it(`should return correct permissions for ${user} users`, () => {
      expect(checkEditPermission(accessUsers[user])).toMatchSnapshot()
    })
  }
  it('should return 1 for a user with admin and edit in diff ns', () => {
    expect(checkEditPermission(mixedUsers.mixedNsAdminEdit)).toMatchSnapshot()
  })
  it('should return 1 for a user with edit and view in diff ns', () => {
    expect(checkEditPermission(mixedUsers.mixedNsEditView)).toMatchSnapshot()
  })
})

describe('checkEditRole', () => {
  for (const user in accessUsers) {
    it(`should return correct permissions for ${user} users`, () => {
      expect(checkEditRole(accessUsers[user][0].rules)).toMatchSnapshot()
    })
  }
  it('should return 1 for a user with admin and edit in diff kinds', () => {
    expect(checkEditRole(mixedUsers.mixedKindAdminEdit[0].rules, extraPolicyKeys)).toMatchSnapshot()
  })
  it('should return 0 for a user with edit and view in diff kinds', () => {
    expect(checkEditRole(mixedUsers.mixedKindEditView[0].rules, extraPolicyKeys)).toMatchSnapshot()
  })
})
