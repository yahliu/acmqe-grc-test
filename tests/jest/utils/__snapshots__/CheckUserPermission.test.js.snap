// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`checkCreatePermission should return 0 for a user with edit and view in diff ns 1`] = `0`;

exports[`checkCreatePermission should return 1 for a user with admin and edit in diff ns 1`] = `1`;

exports[`checkCreatePermission should return correct permissions for admin users 1`] = `1`;

exports[`checkCreatePermission should return correct permissions for clusterAdmin users 1`] = `1`;

exports[`checkCreatePermission should return correct permissions for edit users 1`] = `0`;

exports[`checkCreatePermission should return correct permissions for view users 1`] = `0`;

exports[`checkCreateRole should return 0 for a user with admin and edit in diff kinds 1`] = `0`;

exports[`checkCreateRole should return 0 for a user with edit and view in diff kinds 1`] = `0`;

exports[`checkCreateRole should return correct permissions for admin users 1`] = `1`;

exports[`checkCreateRole should return correct permissions for clusterAdmin users 1`] = `1`;

exports[`checkCreateRole should return correct permissions for edit users 1`] = `0`;

exports[`checkCreateRole should return correct permissions for view users 1`] = `0`;

exports[`checkEditPermission should return 1 for a user with admin and edit in diff ns 1`] = `1`;

exports[`checkEditPermission should return 1 for a user with edit and view in diff ns 1`] = `1`;

exports[`checkEditPermission should return correct permissions for admin users 1`] = `1`;

exports[`checkEditPermission should return correct permissions for clusterAdmin users 1`] = `1`;

exports[`checkEditPermission should return correct permissions for edit users 1`] = `1`;

exports[`checkEditPermission should return correct permissions for view users 1`] = `0`;

exports[`checkEditRole should return 0 for a user with edit and view in diff kinds 1`] = `0`;

exports[`checkEditRole should return 1 for a user with admin and edit in diff kinds 1`] = `1`;

exports[`checkEditRole should return correct permissions for admin users 1`] = `1`;

exports[`checkEditRole should return correct permissions for clusterAdmin users 1`] = `1`;

exports[`checkEditRole should return correct permissions for edit users 1`] = `1`;

exports[`checkEditRole should return correct permissions for view users 1`] = `0`;
